### DISCUSSION TITLE
conversion from cuda(.cu) to spirv(.spv) with chipshar

### DISCUSSION THREAD

**Discussion by zzj-github** [CHIP-SPV/chipStar#755] [General]
I want to implement the conversion from cuda(.cu) to spirv(.spv) with chipstar.
The following are the conversion methods provided in the documentation:

Compiling CUDA applications directly with chipStar
Compilation of CUDA sources without changing the sources, can be done in two ways: The first way is to replace calls of the nvcc compiler with calls of the wrapper script /bin/cuspv in Makefiles. The wrapper script will call clang with the correct flags. The other way is by using CMake: use and then use . However, the project must be compiled with a Clang version supported by HIP. Note that it's not necessary to have Nvidia's CUDA installed.find_package(HIP REQUIRED CONFIG)target_link_libraries(<YOUR_TARGET> hip::device)

The description in the document is too brief and does not provide specific steps, can you describe in detail how to implement these two methods?

**Comment by pjaaskel** [comment]
chipStar can be used only for .cu (sources) to a fat binary which includes .spv as the device program format, not for PTX to SPV (at least so far).

**Comment by zzj-github** [comment]
> chipStar can be used only for .cu (sources) to a fat binary which includes .spv as the device program format, not for PTX to SPV (at least so far).

yes，I want to implement the conversion  from .cu (sources) to a fat binary which includes .spv, how to achieve it? What is the compile command?

**Comment by linehill** [comment]
There is `cuspvc` tool for compiling CUDA sources. For example,
```
$ cd chipStar/samples/cuda_samples/0_Simple/vectorAdd
$ <chipStar-install-dir>/bin/cuspvc -I../../Common vectorAdd.cu -o foo
```

**Comment by zzj-github** [comment]
> There is `cuspvc` tool for compiling CUDA sources. For example,
> 
> ```
> $ cd chipStar/samples/cuda_samples/0_Simple/vectorAdd
> $ <chipStar-install-dir>/bin/cuspvc -I../../Common vectorAdd.cu -o foo
> ```

the output foo is a binary or a spirv file? if foo is a binary, how to obtain the corresponding spirv file?

**Comment by linehill** [comment]
> the output foo is a binary or a spirv file? if foo is a binary, how to obtain the corresponding spirv file?

Binary. To obtain only the device code, use:
```
$ <chipStar-install-dir>/bin/cuspvc -I../../Common --cuda-device-only -c vectorAdd.cu -o foo.spv.cob
$ clang-offload-bundler --type=o --targets=hip-spirv64-generic --input=foo.spv.cob --unbundle --output foo.spv
```

**Comment by zzj-github** [comment]
> > the output foo is a binary or a spirv file? if foo is a binary, how to obtain the corresponding spirv file?
> 
> Binary. To obtain only the device code, use:
> 
> ```
> $ <chipStar-install-dir>/bin/cuspvc -I../../Common --cuda-device-only -c vectorAdd.cu -o foo.spv.cob
> $ clang-offload-bundler --type=o --targets=hip-spirv64-generic --input=foo.spv.cob --unbundle --output foo.spv
> ```

Thanks
Follow the command you provide can generate the spv file.
I couldn't find anything online about these commands. Where are the instructions for these commands? Could you please provide some relevant materials(documents, web links, etc.) about cuda to spirv for me to learn?

**Comment by linehill** [comment]
> Where are the instructions for these commands?

I mostly learned them by studying Clang’s internal [tests](https://github.com/llvm/llvm-project/tree/main/clang/test) plus trial and error. Clang’s [documentation](https://clang.llvm.org/docs/index.html) may be helpful too.

**Comment by zzj-github** [comment]
> > Where are the instructions for these commands?
> 
> I mostly learned them by studying Clang’s internal [tests](https://github.com/llvm/llvm-project/tree/main/clang/test) plus trial and error. Clang’s [documentation](https://clang.llvm.org/docs/index.html) may be helpful too.

Thanks.

**Comment by yuygfgg** [comment]
I don't see cuspvc in my docker installation. Is there a replacement or I made an mistake?

**Reply by linehill** [reply]
The `cuspvc` has been removed and replaced with `cucc`. You can find more details about the new tool [here](https://github.com/CHIP-SPV/chipStar/blob/main/docs/Using.md#compiling-cuda-applications-directly-with-chipstar).
